### PR TITLE
Add `--target` support to `sync` and `install`

### LINKED ISSUE #1517: Add `-t/--target` flag for `uv pip install`
This option is used a lot for vendoring and for custom build systems. 

From `pip install --help`:
```sh
  -t, --target <dir>          Install packages into <dir>. By
                              default this will not replace
                              existing files/folders in <dir>. Use
                              --upgrade to replace existing
                              packages in <dir> with new versions.
```

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

The approach taken here is to model `--target` as an install scheme in which all the directories are just subdirectories of the `--target`. From there, everything else... just works? Like, upgrade, uninstalls, editables, etc. all "just work".

Closes #1517.

### CODE REVIEW

File: crates/uv/tests/pip_sync.rs [hunk 1]
```diff
@@ -4814,3 +4814,72 @@ fn require_hashes_registry_invalid_hash() -> Result<()> {
...
+    ----- stderr -----
+    Resolved 1 package in [TIME]
+    Downloaded 1 package in [TIME]
+    Uninstalled 1 package in [TIME]
+    Installed 1 package in [TIME]
+     - anyio==4.2.0
+     + flask==3.0.3
+    "###);
+
+    Ok(())
```
Comment (zanieb) [inline crates/uv/tests/pip_sync.rs:4884]: Can we include some sort of file system assertions here?

Comment (charliermarsh) [reply crates/uv/tests/pip_sync.rs:4884]: E.g., that the packages are inside the target directory?

Comment (zanieb) [reply crates/uv/tests/pip_sync.rs:4907]: Yeah

**Review by konstin — APPROVED** [review]

**Comment by zanieb** [comment]
This is different than pip right? Want to note this in the compatibility doc?

**Comment by charliermarsh** [comment]
Hmm, I don't think the _user experience_ actually differs from pip in significant ways, or at least in ways that would work in pip but wouldn't work in uv.

**Comment by henryiii** [comment]
Hmm, I tried this out:

```console
$ cargo run --release --bin uv -- pip install importlib_metadata --target tmpdir
$ ls
bin/                                importlib_metadata-7.1.0.dist-info/ zipp/
importlib_metadata/                 include/                            zipp-3.18.1.dist-info/
```

This looks close to what I'd expect, though the `bin` and `include` (empty) directories were unexpected. I would not expect to be able to `import bin` or `import include` :).

**Comment by charliermarsh** [comment]
Yeah, but if you `pip install black --target tmpdir`, you will get a `tmpdir/bin` that contains `black` and `blackd` -- this follows pip. We're just creating those directories upfront rather than lazily. Does have a material impact?

**Comment by charliermarsh** [comment]
> At least for the use case of building zipapps, which is what I use --target for, and seems to be a popular usage, you only want platlib/purelib (in the same folder) and don't want data/scripts/headers, which can't be used from a zipapp.

Yes, but pip adds all of these, as far as I can tell? It just adds them to the top-level. If you `pip install jupyter --target foo`, you will see `foo/share` which contains a bunch of icons and such for Jupyter.

**Comment by charliermarsh** [comment]
I understand that you don't need these things for a zipapp. What I'm trying to understand is why it's a problem to include them, if it matches pip's behavior.

**Comment by henryiii** [comment]
I think it's just surprising that these were here; I've not bundled a library with scripts or headers in a zipapp before so have never seen them and didn't know pip would stick them there.

**Comment by charliermarsh** [comment]
Perhaps I can change it to create those directories lazily so it's less surprising (since they typically would not be created at all in that case).

**Comment by charliermarsh** [comment]
Okay, https://github.com/astral-sh/uv/pull/3274 should give you something closer to what you're used to :)

**Comment by henryiii** [comment]
Nice, that feels a lot more pip-like. :)

**Comment by Pixel-Minions** [comment]
Works great! Thank you.

**Comment by mpderbec** [comment]
Just tried it out... With `--target` support, I wouldn't expect to have to create a virtual environment, correct?

(I'm getting this error:)

```
C:\WINDOWS\system32\cmd.exe /c "uv pip install --target="C:\GitHub\crab\site-packages\x64\prod" -r "C:\GitHub\crab\requirements.prod.txt" "
error: Failed to locate a virtualenv or Conda environment (checked: `VIRTUAL_ENV`, `CONDA_PREFIX`, and `.venv`). Run `uv venv` to create a virtualenv.
```

**Comment by charliermarsh** [comment]
Mmm, I could see arguments either way given our stance on environments, but probably shouldn't be required, no. You can run with `--system` in the meantime.

**Comment by Pixel-Minions** [comment]
You can use --python and pass thr path to the interpreter, it works great
even with embedded versions of python with no venv

**Comment by mpderbec** [comment]
`--system` works great!

**Pull request merged**